### PR TITLE
Add Hellinger Distance for univariate drift detection on continuous data 

### PR DESCRIPTION
**Pull request by Dbhasin1**
- [x] Add reference implementation for Hellinger 
- [x] Add tests 
- [x] Add documentation 
- [ ] Add comparison to other methods

### CODE REVIEW

**Comment by nikml** [comment]
Let's not merge this before we complete the new release.

**Comment by jakubnml** [comment]
@Dbhasin1 

Double-check the hellinger formula. I think you have a square of sum while you should have a sum of squares.

**Comment by Dbhasin1** [comment]
> @Dbhasin1
> 
> Double-check the hellinger formula. I think you have a square of sum while you should have a sum of squares.

Resolved.

File: .zshrc [hunk 1]
```diff
@@ -0,0 +1 @@
+
```
Comment (nikml) [inline .zshrc:1]: Is this file needed? Looks like a local shell file, not something needed for the repository

File: docs/how_it_works/univariate_drift_detection.rst [hunk 2]
```diff
@@ -108,6 +108,36 @@ Wasserstein distance between the two distributions is given by:
 .. math::
     W_1\left(X_i^{ref},X_i^{ana}\right) = \int_\mathbb{R}\left|\hat{F}_{ref}(x)-\hat{F}_{ana}(x)\right|dx
 
+.. _univariate-drift-detection-cont-hellinger:
+
+Hellinger Distance
+........................
+
+The Hellinger Distance, is a distance metric used to quantify the similarity between two distributions. It measures the overlap between the probabilities assigned 
+to the same event by both reference and analysis. It ranges from 0 to 1 where a value of 1 is only achieved when reference assigns zero probability to each event to which
```
Comment (nikml) [inline docs/how_it_works/univariate_drift_detection.rst:117]: It would be good to say that reference and analysis are samples. Also - and this likely applies for other methods as well - I m not too keen on naming them reference and analysis. Because while we may be using the whole reference dataset on one side, we are only using a chunk on the other. Speaking about analysis here may confuse people.

File: nannyml/drift/univariate/methods.py [hunk 3]
```diff
@@ -481,3 +481,80 @@ def _alert(self, data: pd.Series):
...
+            len_data = len(data)
+            data_proba_in_bins = np.histogram(data, bins=self._bins)[0] / len_data
+
+        else:
+            data_unique, data_counts = np.unique(data, return_counts=True)
+            data_counts_dic = dict(zip(data_unique, data_counts))
+            data_count_on_ref_bins = [data_counts_dic[key] if key in data_counts_dic else 0 for key in self._bins]
+            data_proba_in_bins = np.array(data_count_on_ref_bins) / len(data)
+
+        leftover = 1 - np.sum(data_proba_in_bins)
```
Comment (nikml) [inline nannyml/drift/univariate/methods.py:544]: It would be good to add a comment here about what leftover is and why we need it.

Also, assuming I m getting this right @jakubnml are leftovers always from one side, or could the be from both (before start and after finish). Or this doesn't matter?

File: docs/how_it_works/univariate_drift_detection.rst [hunk 4]
```diff
@@ -192,6 +222,21 @@ To help our intuition we can look at the image below:
 We see how the relative frequencies of three categories have changed between reference and analysis data.
 We also see that the JS Divergence contribution of each change and the resulting JS distance.
 
+.. _univ_cat_method_hellinger:
+
+Hellinger Distance
+........................
+
+The Hellinger Distance, is a distance metric used to quantify the similarity between two distributions. It measures the overlap between the probabilities assigned
```
Comment (nikml) [inline docs/how_it_works/univariate_drift_detection.rst:230]:
```suggestion
The Hellinger Distance, is a distance metric used to quantify the similarity between two probability distributions. It measures the overlap between the probabilities assigned 
```

File: docs/how_it_works/univariate_drift_detection.rst [hunk 5]
```diff
@@ -192,6 +222,21 @@ To help our intuition we can look at the image below:
 We see how the relative frequencies of three categories have changed between reference and analysis data.
 We also see that the JS Divergence contribution of each change and the resulting JS distance.
 
+.. _univ_cat_method_hellinger:
+
+Hellinger Distance
+........................
+
+The Hellinger Distance, is a distance metric used to quantify the similarity between two distributions. It measures the overlap between the probabilities assigned 
+to the same event by both reference and analysis. It ranges from 0 to 1 where a value of 1 is only achieved when reference assigns zero probability to each event to which
```
Comment (nikml) [inline docs/how_it_works/univariate_drift_detection.rst:231]:
```suggestion
to the same event by both reference and analysis samples. It ranges from 0 to 1 where a value of 1 is only achieved when reference assigns zero probability to each event to which 
```

File: docs/how_it_works/univariate_drift_detection.rst [hunk 6]
```diff
@@ -192,6 +222,21 @@ To help our intuition we can look at the image below:
...
 
+.. _univ_cat_method_hellinger:
+
+Hellinger Distance
+........................
+
+The Hellinger Distance, is a distance metric used to quantify the similarity between two distributions. It measures the overlap between the probabilities assigned 
+to the same event by both reference and analysis. It ranges from 0 to 1 where a value of 1 is only achieved when reference assigns zero probability to each event to which 
+analysis assigns some positive probability and vice versa. 
+The formula is given by:
```
Comment (nikml) [inline docs/how_it_works/univariate_drift_detection.rst:233]:
```suggestion
```

File: docs/how_it_works/univariate_drift_detection.rst [hunk 7]
```diff
@@ -192,6 +222,21 @@ To help our intuition we can look at the image below:
...
+
+For a categorical feature Hellinger Distance is defined as:
+
+.. math::
+ H\left(X_i^{ref},X_i^{ana}\right) = \frac{1}{\sqrt{2}}\left[\sum_{x \in X}\left(\sqrt{{F}_{ref}(x)}-\sqrt{{F}_{ana}(x)}\right)^2dx\right]^{1/2}
```
Comment (nikml) [inline docs/how_it_works/univariate_drift_detection.rst:238]:
```suggestion
 H\left(X_i^{ref},X_i^{ana}\right) = \frac{1}{\sqrt{2}}\left[\sum_{x \in X}\left(\sqrt{{F}_{ref}(x)}-\sqrt{{F}_{ana}(x)}\right)^2\right]^{1/2}
```
Integral ``dx`` leftover from formula here.

File: docs/how_it_works/univariate_drift_detection.rst [hunk 8]
```diff
@@ -192,6 +222,21 @@ To help our intuition we can look at the image below:
...
+
```
Comment (nikml) [inline docs/how_it_works/univariate_drift_detection.rst:239]: It would also be good to explain that F in the categorical case correspond to Probability Mass Functions.

File: docs/how_it_works/univariate_drift_detection.rst (hunk 4 above)
Comment (nikml) [inline docs/how_it_works/univariate_drift_detection.rst:230]: Make `Hellinger Distance` a link to the relevant Wiki page?

File: docs/how_it_works/univariate_drift_detection.rst [hunk 9]
```diff
@@ -108,6 +108,36 @@ Wasserstein distance between the two distributions is given by:
...
+.. math::
+    H\left(X_i^{ref},X_i^{ana}\right) = \frac{1}{\sqrt{2}}\left[\int_{}\left(\sqrt{{F}_{ref}(x)}-\sqrt{{F}_{ana}(x)}\right)^2dx\right]^{1/2}
+
+The implementation splits a continuous feature into bins, calculates the relative frequency
+for each bin from reference and analyzed data and calculates the
+resulting Hellinger Distance. The binning is done using `Doane's formula`_ from numpy.
+If a continuous feature has relatively low amount of unique values, meaning that
+unique values are less then 10% of the reference dataset size up to a maximum of 50, each value becomes a bin.
+
+This distance is very closely related to the Bhattacharya Coefficient. However we choose the former because it follows the triangle inequality and is
```
Comment (nikml) [inline docs/how_it_works/univariate_drift_detection.rst:130]: Since in the implementation we are not using BC I m not sure we want have this in our docs.

Thoughts @jakubnml ?

File: docs/how_it_works/univariate_drift_detection.rst [hunk 10]
```diff
@@ -108,6 +108,36 @@ Wasserstein distance between the two distributions is given by:
...
+
+The Hellinger Distance, is a distance metric used to quantify the similarity between two distributions. It measures the overlap between the probabilities assigned 
+to the same event by both reference and analysis. It ranges from 0 to 1 where a value of 1 is only achieved when reference assigns zero probability to each event to which 
+analysis assigns some positive probability and vice versa. 
+The formula is given by:
+
+.. math::
+    H\left(X_i^{ref},X_i^{ana}\right) = \frac{1}{\sqrt{2}}\left[\int_{}\left(\sqrt{{F}_{ref}(x)}-\sqrt{{F}_{ana}(x)}\right)^2dx\right]^{1/2}
+
+The implementation splits a continuous feature into bins, calculates the relative frequency
```
Comment (nikml) [inline docs/how_it_works/univariate_drift_detection.rst:124]: IT would be good to improve the language a bit here, for example:
```suggestion
In order to Calculate Hellinger Distance NannyML splits a continuous feature into bins, calculates the relative frequency
```

**Review by nikml — CHANGES_REQUESTED** [review]

**Comment by nnansters** [comment]
@Dbhasin1 there is a test failing.

```python
def test_hellinger_both_continuous_analysis_with_small_drift():
    np.random.seed(1)
    reference = pd.Series(np.random.normal(0, 1, 10_000))
    analysis = pd.Series(np.random.normal(-2, 1, 10_000))
    hell_dist = HellingerDistance().fit(reference).calculate(analysis)
    hell_dist = np.round(hell_dist, 2)
    assert hell_dist == 0.62  # fails because the actual is 0.63
```

Can you confirm/correct?

File: docs/how_it_works/univariate_drift_detection.rst (hunk 2 above)
Comment (jakubnml) [reply docs/how_it_works/univariate_drift_detection.rst:117]: Damn, good point. Maybe in how it works where, we describe all the methods, we should just name them p and q, or sample A and B, or whatever. Apart from that we should have a paragraph that explains how this works in the library, that reference set serves as sample A, and every chunk of data that is inputed to _calculate/estimate_ method is the B sample?

File: nannyml/drift/univariate/methods.py (hunk 3 above)
Comment (jakubnml) [reply nannyml/drift/univariate/methods.py:541]: For continuous data could be from both sides. For categorical data leftovers is all the unseen categories. According to the experiments we did, for both of the methods it doesn't matter if you create multiple _bins_ for chunk data and multiple _bins_ with for reference data (all with 0 relative frequency). For other methods (that we dont have in NML yet) it may be important.

Comment (jakubnml) [reply nannyml/drift/univariate/methods.py:541]: By both methods I mean both methods that use the same binning logic - JS and Hell.

File: .zshrc (hunk 1 above)
Comment (Dbhasin1) [reply .zshrc:1]: yeah dont know where this came from, will remove in the next commit

**Comment by Dbhasin1** [comment]
> @Dbhasin1 there is a test failing.
> 
> ```python
> def test_hellinger_both_continuous_analysis_with_small_drift():
>     np.random.seed(1)
>     reference = pd.Series(np.random.normal(0, 1, 10_000))
>     analysis = pd.Series(np.random.normal(-2, 1, 10_000))
>     hell_dist = HellingerDistance().fit(reference).calculate(analysis)
>     hell_dist = np.round(hell_dist, 2)
>     assert hell_dist == 0.62  # fails because the actual is 0.63
> ```
> 
> Can you confirm/correct?

Yeah, this happened because I didn't take the `np.random.seed(1)` into account. I had calculated my result without it which was bound to be different. Resolving this
